lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2472,14 +2472,48 @@ namespace {
         ctors.push_back(valueCtor);
       }
 
-      // Do not allow Swift to construct foreign reference types (at least, not
-      // yet).
       if (isa<StructDecl>(result)) {
         for (auto ctor : ctors) {
           // Add ctors directly as they cannot always be looked up from the
           // clang decl (some are synthesized by Swift).
           result->addMember(ctor);
         }
+      } else {
+        if (Impl.SwiftContext.LangOpts.hasFeature(
+                Feature::CXXForeignReferenceTypeInitializers)) {
+          assert(isa<ClassDecl>(result) && "Expected result to be a ClassDecl "
+                                           "for C/C++ foreign reference types");
+          // When we add full support for C foreign reference types then we
+          // should synthesize static factories for them as well
+          if (auto *cxxRecordDecl = const_cast<clang::CXXRecordDecl *>(
+                  dyn_cast<clang::CXXRecordDecl>(decl))) {
+            bool hasUserProvidedStaticFactory = false;
+            for (const auto *method : cxxRecordDecl->methods()) {
+              if (method->isStatic()) {
+                for (const auto *attr :
+                     method->specific_attrs<clang::SwiftNameAttr>()) {
+                  std::string attributeName = attr->getName().str();
+                  if (attributeName.find("init") != std::string::npos) {
+                    hasUserProvidedStaticFactory = true;
+                    break;
+                  }
+                }
+              }
+              if (hasUserProvidedStaticFactory)
+                break;
+            }
+            if (!hasUserProvidedStaticFactory) {
+              if (auto generatedCxxMethodDecl = SwiftDeclSynthesizer::
+                      synthesizeStaticFactoryForCXXForeignRef(
+                          cxxRecordDecl, cxxRecordDecl->getASTContext(),
+                          Impl.getClangSema())) {
+                if (Decl *importedInitDecl =
+                        VisitCXXMethodDecl(generatedCxxMethodDecl))
```

Comment:
We should be calling into `importDecl`/`importDeclDirectly` here instead of `VisitCXXMethodDecl`, since the former will handle caching, forward declarations, circular references, etc.